pypeman/plugins/base.py

```diff
@@ -0,0 +1,137 @@
+"""
+Base class for a pypeman plugin
+
+All interfaces are specified nothing is implemented
+"""
+
+import logging
+
+from inspect import iscoroutinefunction
+
+
+logger = logging.getLogger(__name__)
+
+
+class BasePlugin:
+    (INITIALIZED, READY, STARTING, STARTED, STOPPING, STOPPED, DESTROYED) = \
+        states = range(7)
+
+    STATE_NAMES = (
+        "INITIALIZED", "READY", "STARTING", "STARTED",
+        "STOPPING", "STOPPED", "DESTROYED",
+        )
+
+    def __init__(self):
+        """
+        Early initialisation before any pypeman endpoints and channels are created
+        """
+        self.name = __name__ + str(self.__class__)
+        logger.debug("Initialized plugin %s", self.name)
+        self.do_init()
+        self.status = self.INITIALIZED
+        self.loop = None
+        self.to_start = []
+        self.start_results = []
+
+    def do_init(self):
+        pass
+
+    def ready(self):
+        """
+        Code to be executed after the whole pypeman graph has been generated.
+        ready() functions of all plugins are executed before the loop is generated
+        """
+        logger.debug("called ready for plugin %s", self.name)
+        assert self.status == self.INITIALIZED
+        self.do_ready()
+        self.status = self.READY
+
+    def do_ready(self):
+        pass
+
+    def set_loop(self, loop):
+        self.loop = loop
+
+    def set_starting(self):
+        logger.debug("starting plugin %s", self.name)
+        assert self.status == self.READY
+        self.status = self.STARTING
+
+    def set_started(self):
+        self.status = self.STARTED
+
+    async def start(self):
+        """
+        Code to be executed after the creation of the event loop
+        """
+        self.set_starting()
+        if iscoroutinefunction(self.do_start):
+            logger.debug("awaiting coro do_start()")
+            future = await self.do_start()
+            logger.debug("started => %s", repr(future))
+            self.start_results.append(future)
+        else:
+            start_rslt = self.do_start()
+            logger.debug("do_start rslt = %s", repr(start_rslt))
+            if isinstance(start_rslt, (list, tuple)):
+                logger.debug("isnocoro")
+                for coro in start_rslt:
+                    rslt = await coro
+                    self.start_results.append(rslt)
+            elif start_rslt is None:
+                pass
+                logger.debug("None")
```

Comment:
Debug message too small or too long ;-)

Reply:
will do

Reply:
Done. to be committed with some more changes to be done.